### PR TITLE
Add missing tests for app.js

### PR DESCRIPTION
**Pull request by hatched**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6698/

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -1838,7 +1838,7 @@ YUI.add('juju-gui', function(Y) {
           this.state.changeState({
             root: null,
             store: null,
-            model:null,
+            model: null,
```
Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: Thanks.

File: jujugui/static/gui/src/test/test_app.js [hunk 2]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+        jujuCoreVersion: '2.0.0'
+      });
+    });
+
+    afterEach(function() {
+      app.destroy();
+    });
+
+    it('calls next and returns if root state is new', () => {
+      const next = sinon.stub();
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: `const next = sinon.stub().withArgs()`?

File: jujugui/static/gui/src/test/test_app.js [hunk 3]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+        viewContainer: container,
+        jujuCoreVersion: '2.0.0'
+      });
+    });
+
+    afterEach(function() {
+      app.destroy();
+    });
+
+    it('properly handles a resolved entity promise', done => {
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: s/resolved/rejected/ ?

File: jujugui/static/gui/src/test/test_app.js [hunk 4]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+      const entityPromise = Promise.reject(userState);
+      app.controllerAPI.userIsAuthenticated = true;
+      app._listAndSwitchModel = args => {
+        assert.deepEqual(args, userState);
+        done();
+      };
+      app._disambiguateUserState(entityPromise);
+    });
+
+    it('properly handles a rejected entity promise', done => {
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: s/rejected/resolved/ ;-)

File: jujugui/static/gui/src/test/test_app.js [hunk 5]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+        'lastConnection':null
+      }];
+    });
+
+    afterEach(function() {
+      app.destroy();
+    });
+
+    it('switches to a supplied model path', () => {
+      app.maskVisibility = sinon.stub();
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: ISTM that these initial lines are the same for each tests, so I guess they can be moved to beforeEach:
```
app.maskVisibility = sinon.stub();
app.state.changeState = sinon.stub();
app._getAuth = sinon.stub().returns({rootUserName: 'pug'});
app.controllerAPI.listModelsWithInfo = function(callback) {
  callback.call(app, null, modelList);
};
```

Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: Also, maybe `app.maskVisibility = sinon.stub().withArgs(false);`?

File: jujugui/static/gui/src/test/test_app.js [hunk 6]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+      app.state.changeState = sinon.stub();
+      app._getAuth = sinon.stub().returns({rootUserName: 'pug'});
+      app.controllerAPI.listModelsWithInfo = function(callback) {
+        callback.call(app, null, modelList);
+      };
+      app._listAndSwitchModel('frankban/latta');
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: I'd include the second uuid argument as null. For sanity.

File: jujugui/static/gui/src/test/test_app.js [hunk 7]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+    });
+
+    it('switches to disconnected state if no model found', () => {
+      app.maskVisibility = sinon.stub();
+      app.state.changeState = sinon.stub();
+      app._getAuth = sinon.stub().returns({rootUserName: 'pug'});
+      app.controllerAPI.listModelsWithInfo = function(callback) {
+        callback.call(app, null, modelList);
+      };
+      app._listAndSwitchModel(null, 'bad-uuid');
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: Let's do the same "not found" check with a model path please.

File: jujugui/static/gui/src/test/test_app.js [hunk 8]
```diff
@@ -1954,7 +1954,396 @@ describe('App', function() {
...
+      assert.equal(app.maskVisibility.callCount, 1);
+      assert.equal(app.state.changeState.callCount, 1);
+      assert.deepEqual(app.state.changeState.args[0], [{
+        root: null, store: null, model: null, user: null,
+        profile: 'pug'
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_app.js]: // TODO: check that the user is notified about the model not being found?

File: jujugui/static/gui/src/test/test_app.js (hunk 2 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: next isn't called with args

File: jujugui/static/gui/src/test/test_app.js (hunk 3 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: Heh oops, fixed

File: jujugui/static/gui/src/test/test_app.js (hunk 4 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: and fixed

File: jujugui/static/gui/src/test/test_app.js (hunk 5 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: Good call, done

Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: assertians added for maskVisibility

File: jujugui/static/gui/src/test/test_app.js (hunk 6 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: done

File: jujugui/static/gui/src/test/test_app.js (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/test/test_app.js]: test added

**Comment by hatched** [comment]
Thanks all for the reviews! :shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6700/

**Pull request merged**